# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

1. Create a new DataFrame that only includes customers who:
   - have a **low total_claim_amount** (e.g., below $1,000),
   - have a response "Yes" to the last marketing campaign.

2. Using the original Dataframe, analyze:
   - the average `monthly_premium` and/or customer lifetime value by `policy_type` and `gender` for customers who responded "Yes", and
   - compare these insights to `total_claim_amount` patterns, and discuss which segments appear most profitable or low-risk for the company.

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

In [42]:
# your code goes here
# Data Inspection Script
# url: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
import warnings
warnings.filterwarnings('ignore')
df = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv')
print(df.head())
print(df.shape)
print(df.info())
print(df.describe())
print(df.isnull().sum())
print(df.columns)
print(df.dtypes)
print(df.nunique())

# print unique values for each column
for col in df.columns:
    print(f"\nUnique values in column {col}: {df[col].unique()}")




   Unnamed: 0 Customer       State  Customer Lifetime Value Response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   Coverage Education Effective To Date EmploymentStatus Gender  ...  \
0     Basic   College           2/18/11         Employed      M  ...   
1     Basic   College           1/18/11       Unemployed      F  ...   
2     Basic  Bachelor           2/10/11         Employed      M  ...   
3  Extended   College           1/11/11         Employed      M  ...   
4   Premium  Bachelor           1/17/11    Medical Leave      F  ...   

   Number of Open Complaints Number of Policies     Policy Type        Policy  \
0                        0.0                  9  Corp

In [43]:
# Data Cleaning
# -----------------------
# Drop Unnamed: 0
# Standardize column names

df = df.drop(columns=["Unnamed: 0"], errors="ignore")

df.columns = (
    df.columns
      .str.lower()
      .str.replace(" ", "_", regex=False)
)

# Quick check
print(df.shape)
df.head()


(10910, 25)


customer       state  customer_lifetime_value response  coverage education  \
0  DK49336     Arizona              4809.216960       No     Basic   College   
1  KX64629  California              2228.525238       No     Basic   College   
2  LZ68649  Washington             14947.917300       No     Basic  Bachelor   
3  XL78013      Oregon             22332.439460      Yes  Extended   College   
4  QA50777      Oregon              9025.067525       No   Premium  Bachelor   

  effective_to_date employmentstatus gender  income  ...  \
0           2/18/11         Employed      M   48029  ...   
1           1/18/11       Unemployed      F       0  ...   
2           2/10/11         Employed      M   22139  ...   
3           1/11/11         Employed      M   49078  ...   
4           1/17/11    Medical Leave      F   23675  ...   

  number_of_open_complaints number_of_policies     policy_type        policy  \
0                       0.0                  9  Corporate Auto  Corporate L3   
1                       0.0                  1   Personal Auto   Personal L3   
2                       0.0                  2   Personal Auto   Personal L3   
3                       0.0                  2  Corporate Auto  Corporate L3   
4                       NaN                  7   Personal Auto   Personal L2   

   renew_offer_type  sales_channel  total_claim_amount  vehicle_class  \
0            Offer3          Agent          292.800000  Four-Door Car   
1            Offer4    Call Center          744.924331  Four-Door Car   
2            Offer3    Call Center          480.000000            SUV   
3            Offer2         Branch          484.013411  Four-Door Car   
4            Offer1         Branch          707.925645  Four-Door Car   

  vehicle_size vehicle_type  
0      Medsize          NaN  
1      Medsize          NaN  
2      Medsize            A  
3      Medsize            A  
4      Medsize          NaN  

[5 rows x 25 columns]

In [44]:
df["number_of_open_complaints"].dtype


dtype('float64')

At this stage, the dataset has appropriate data types and standardized categorical values to proceed with filtering and aggregation tasks.
>
Minor inconsistencies (e.g., NaN values or float representation of count data) do not affect the current analysis.



In [45]:
# Task 1: Filter: low total_claim_amount + response "Yes"
# Create a new DataFrame with customers who:
# - Have low total_claim_amount (below 1000)
# - Responded "Yes" to the last marketing campaign


filtered_df = df[
    (df["total_claim_amount"] < 1000) &
    (df["response"] == "Yes")
]

filtered_df.head()
filtered_df.shape

print(filtered_df.shape)
print(filtered_df.head())

# save filtered_df to a csv file
filtered_df.to_csv("filtered_customers.csv", index=False)

# Filter customers who responded "Yes" and have total_claim_amount below 1000 (low-claim responders)


(1399, 25)
   customer       state  customer_lifetime_value response  coverage education  \
3   XL78013      Oregon             22332.439460      Yes  Extended   College   
8   FM55990  California              5989.773931      Yes   Premium   College   
15  CW49887  California              4626.801093      Yes     Basic    Master   
19  NJ54277  California              3746.751625      Yes  Extended   College   
27  MQ68407      Oregon              4376.363592      Yes   Premium  Bachelor   

   effective_to_date employmentstatus gender  income  ...  \
3            1/11/11         Employed      M   49078  ...   
8            1/19/11         Employed      M   66839  ...   
15           1/16/11         Employed      F   79487  ...   
19           2/26/11         Employed      F   41479  ...   
27           2/28/11         Employed      F   63774  ...   

   number_of_open_complaints number_of_policies     policy_type        policy  \
3                        0.0                  2  Corpo

In [46]:
# Task 2: Analyze average monthly premium and/or CLV by policy_type & gender (Response == "Yes") + compare with claims
# Using the original Dataframe, analyze:
# - the average `monthly_premium` and/or customer lifetime value by `policy_type` and `gender` for customers who responded "Yes", and
# - compare these insights to `total_claim_amount` patterns, and discuss which segments appear most profitable or low-risk for the company.

yes_responders = df[df["response"] == "Yes"]
grouped_analysis = (
    yes_responders
    .groupby(["policy_type", "gender"], as_index=False)
    .agg({
        "monthly_premium_auto": "mean",
        "customer_lifetime_value": "mean",
        "total_claim_amount": "mean"
    })
    .round(2)
)


print(grouped_analysis.sort_values("customer_lifetime_value", ascending=False))


      policy_type gender  monthly_premium_auto  customer_lifetime_value  \
2   Personal Auto      F                 99.00                  8339.79   
5    Special Auto      M                 86.34                  8247.09   
1  Corporate Auto      M                 92.19                  7944.47   
0  Corporate Auto      F                 94.30                  7712.63   
4    Special Auto      F                 92.31                  7691.58   
3   Personal Auto      M                 91.09                  7448.38   

   total_claim_amount  
2              452.97  
5              429.53  
1              408.58  
0              433.74  
4              453.28  
3              457.01  


In [47]:
# Task 3: Analyzeustomers per state, keep only states with > 500 customers
# Using the original Dataframe, analyze the total number of customers who have policies in each state, 
# and then filter the results to only include states where there are more than 500 customers.

state_customer_counts = df['state'].value_counts()
filtered_states = state_customer_counts[state_customer_counts > 500]
print(filtered_states)

state
California    3552
Oregon        2909
Arizona       1937
Nevada         993
Washington     888
Name: count, dtype: int64


**Insights:**
>
California has the highest number of customers, followed by Oregon and Arizona.
>
All states in the dataset have more than 500 customers, indicating a strong geographic coverage across the analyzed regions.

In [48]:
# Task 4: Analyze Min, Max, Median CLV by education level and gender
# Using the original Dataframe, find the maximum, minimum, and median customer lifetime value 
# by education level and gender. 
# Write your conclusions.
clv_stats = (
    df
    .groupby(["education", "gender"], as_index=False)
    .agg({      
        "customer_lifetime_value": ["min", "max", "median"]
    })
)
clv_stats.columns = ["education_level", "gender", "min_clv", "max_clv", "median_clv"]
print(clv_stats.sort_values("median_clv", ascending=False))         
print(clv_stats.sort_values("median_clv", ascending=False))


        education_level gender      min_clv      max_clv   median_clv
7  High School or Below      M  1940.981221  83325.38119  6286.731006
6  High School or Below      F  2144.921535  55277.44589  6039.553187
3               College      M  1918.119700  61134.68307  6005.847375
8                Master      F  2417.777032  51016.06704  5729.855012
0              Bachelor      F  1904.000852  73225.95652  5640.505303
2               College      F  1898.683686  61850.18803  5623.611187
9                Master      M  2272.307310  50568.25912  5579.099207
5                Doctor      M  2267.604038  32677.34284  5577.669457
1              Bachelor      M  1898.007675  67907.27050  5548.031892
4                Doctor      F  2395.570000  44856.11397  5332.462694
        education_level gender      min_clv      max_clv   median_clv
7  High School or Below      M  1940.981221  83325.38119  6286.731006
6  High School or Below      F  2144.921535  55277.44589  6039.553187
3               Coll

**Insights:**
>
Customers with "High School or Below" education show the highest median CLV for both genders, especially males, indicating strong long-term value despite lower formal education levels.
>
Across most education levels, male customers tend to have slightly higher median CLV than females, although the differences are not large.
>
Doctorate-level customers show comparatively lower median CLV, suggesting that higher education does not necessarily translate into higher customer lifetime value in this dataset.
>
Overall, education level appears to influence CLV, but gender differences are moderate and should be interpreted with caution.

In [ ]:
# Bonus 5:
# Analyze the number of policies sold by state and month. 
# Present the data in a table where the months are arranged as columns and the states are arranged as rows.
# Policies sold by State (rows) and Month (columns)

# Parse the date column
df["effective_to_date"] = pd.to_datetime(df["effective_to_date"], errors="coerce")

# Create a month column (month number)
df["month"] = df["effective_to_date"].dt.month


# Sort months in ascending order
policy_counts = df.pivot_table(
    index="state",
    columns="month",
    aggfunc="size",
    fill_value=0
).sort_index(axis=1)

# show the result in descending order of total policies sold
policy_counts["total_policies"] = policy_counts.sum(axis=1)
policy_counts = policy_counts.sort_values("total_policies", ascending=False)
policy_counts = policy_counts.drop(columns=["total_policies"])


print(policy_counts)

# display the pivot table
policy_counts


month          1     2
state                 
California  1918  1634
Oregon      1565  1344
Arizona     1008   929
Nevada       551   442
Washington   463   425


month          1     2
state                 
California  1918  1634
Oregon      1565  1344
Arizona     1008   929
Nevada       551   442
Washington   463   425

**Insights:**
>
Policy sales are highly concentrated in California and Oregon across both months analyzed.
>
California leads consistently, making it the most important state in terms of sales volume.
>
Arizona also shows solid performance, while Nevada and Washington lag behind.
>
These patterns suggest that sales efforts and marketing resources may be most effective when prioritized in high-volume states, particularly at the beginning of the year.

In [ ]:
# Bonus 6:
# Find top 3 states by total policies sold,
# then show the number of policies sold by month for those states.

total_policies_by_state = (
    df.groupby("state")["policy"]
      .count()
      .sort_values(ascending=False)
)

top3_states = total_policies_by_state.head(3).index
print("\nTop 3 states by policies sold:", list(top3_states))

top3_policies_by_month = (
    df[df["state"].isin(top3_states)]
      .groupby(["state", "month"])["policy"]
      .count()
      .unstack(fill_value=0)
)
# show the result in descending order of total policies sold
top3_policies_by_month["total_policies"] = top3_policies_by_month.sum(axis=1)
top3_policies_by_month = top3_policies_by_month.sort_values("total_policies", ascending=False)
top3_policies_by_month = top3_policies_by_month.drop(columns=["total_policies"])

print("\nPolicies sold by Month for Top 3 states:")
print(top3_policies_by_month)

# display the pivot table
top3_policies_by_month



Top 3 states by policies sold: ['California', 'Oregon', 'Arizona']

Policies sold by Month for Top 3 states:
month          1     2
state                 
California  1918  1634
Oregon      1565  1344
Arizona     1008   929


month          1     2
state                 
California  1918  1634
Oregon      1565  1344
Arizona     1008   929

**Insights:**
>
California, Oregon, and Arizona are the top three states with the highest number of policies sold in both January and February. California leads by a significant margin, followed by Oregon and Arizona, showing consistent sales performance across months.
>
These states represent the strongest markets and should be prioritized for marketing and sales efforts.
>
The analysis is limited to the first two months of the year, so seasonal effects may not be fully captured.

In [ ]:

# Bonus 7:
# Marketing wants the response rate ("Yes") by sales channel.
# Steps:
# - Create a binary flag (1 if Yes else 0)
# - Compute mean of the flag by sales_channel (mean = response rate)

df["response_yes"] = (df["response"] == "Yes").astype(int)

response_rate_by_channel = (
    df.groupby("sales_channel")["response_yes"]
      .mean()
      .mul(100)
      .round(2)
      .sort_values(ascending=False)
)

print("\nResponse rate (%) by Sales Channel:")
print(response_rate_by_channel)



Response rate (%) by Sales Channel:
sales_channel
Agent          18.01
Web            10.89
Branch         10.79
Call Center    10.32
Name: response_yes, dtype: float64


**Insights:**
>
The Agent channel shows the highest response rate (18.01%), significantly outperforming all other sales channels. Web, Branch, and Call Center have similar and considerably lower response rates (around 10–11%).
>
This suggests that personalized, agent-driven interactions are more effective at converting customers in marketing campaigns.
>
Results are based on response rate only and do not account for revenue, costs, or long-term value.